### ISSUE TITLE
[native-image] Add AArch64 support inline objects in runtime-installed code

### ISSUE THREAD

**Issue by teshull** [oracle/graal#2047]
Hi, I'm planning to add support for inline objects within runtime-installed code for AArch64. I thought I would outline my approach here before everything is finished to see what others think.

In AArch64, given that instructions are of fixed 32-bit width, it is not possible to directly inline object pointers into the code. Instead, a series of `mov` instructions must be used to load the entire pointer value into a register. Each `mov` instruction can hold 16-bits of the value. This becomes a problem when storing reference maps about the runtime-installed code. Currently the code reference map encoder/decoder assumes that each reference is stored contiguously. However, this is not possible in AArch64. 

Another problem currently is retrieving/writing pointer values stored in a series of `mov` instructions. Currently, within ReferenceAccess and ObjectReferenceVisitor it also makes the same assumption about values being stored contiguously.

Based on the above information, I propose the following:

1. Introduce the notion of a "split" pointer; that is, a value which is stored
in a sequence of instructions
2. Add new state to the reference map to indicate a reference is split
3. Add a new method to ObjectReferenceVisitor for visiting split references
4. Add two new methods to Reference to read/write split values:

`readSplitObjectAsUntrackedPointer(Pointer p, boolean compressed, int strideSize, int unitSize, int numUnits, int bitOffset)`

`writeSplitObjectAt(Pointer p, Object value, boolean compressed, int strideSize, int unitSize, int numUnits, int bitOffset)`

In this representation, the strideSize is the size of each instruction, the unitSize and numUnits indicate the bit size of each intermediate and number of intermediates to read, and bitOffset is the offset of the beginning of the value within each instruction.

Does this seem like a reasonable approach?

In terms of implementation, how much subclassing should I do for the CodeReference(Encoder/Decoder), ReferenceMaps, and ReferenceVisitors vs. adding extra features to the original classes?

Also, does the notion of a split object make sense, or should I make something more AArch64 specific, considering that currently this is only needed here?

**Comment by tshull226** [comment]
@christianhaeubl please let me know what you think

**Comment by christianhaeubl** [comment]
So, only 16-bit can be encoded in a mov instruction and the GC can move objects. Then, it is necessary to always emit a fixed number of mov instructions for every immediate object pointer:
- compressed references: 2 mov instructions
- uncompressed references: 4 mov instructions

It should be possible to exclude immediate object pointers that point into the image heap from that logic (i.e., the shortest number of mov instructions can be used there as the object won't be moved by the GC and no reference map entry should be necessary neither).

- ad "Add new state to the reference map to indicate a reference is split": this can probably be avoided as all embedded object pointers are split references on AArch64.

On the architecture-independent level (e.g. for the reference maps), it should be possible to keep the concept that references are stored contiguously. In my opinion, there are only two major differences to before:
- the size that an immediate reference occupies in the code is architecture dependent: immediate compressed references have a size of 64 bits on AArch64 (2 mov instructions) and uncompressed references have a size of 128 bits on AArch64 (4 mov instructions).
- in code such as `ObjectReferenceVisitor`, it will be necessary to distinguish between normal object references and object references that are embedded in code so that the correct logic can be used to access the value.

So, I would add an extra boolean to the methods of `ReferenceAccess` to indicate if the accessed reference is embedded in code so that an architecture dependent implementation of `ReferenceAccess` can decide how to deal with those values in detail.

@christianwimmer and @peter-hofer: please also add your thoughts on that topic.

**Comment by tshull226** [comment]
Hi @christianhaeubl,

Thank you for your feedback. One issue is that I believe within runtime installed code not all objects are inlined into the code, but instead there is also a data section containing object pointers at the end which are stored within the same reference map (see RuntimeCodeInstaller.java:278). Within the data section the pointer values are contiguous, so I believe there needs to be a way to differentiate between split and contiguous pointers within the same reference map.

Another point to consider is that currently compressed pointers also appear to be always using 4 instructions (e.g. SubstrateAArch64Backend.java:737), but that may not always be the case in the future.

**Comment by christianhaeubl** [comment]
To distinguish object pointers in the data section from inlined object pointers, we could either:
a. use the address of the reference to detect if it is part of the code or the data section
b. use two different reference maps, one for the references embedded into code, one for the data section.
c. add a new state

I am not a big fan of adding a new state as it will increase the size of all reference maps (I don't think that we have any unused bits there). If you think that there is an efficient way to encode that extra, I am also fine with that solution.

**Comment by tshull226** [comment]
I believe new state can be added without increasing the size of the reference map by leveraging the fact that on AArch64 all pointers will follow at least a 4-byte alignment. This allows for both the compressed and # of instructions (4 vs. 8) to also be encoded within the gap value of the reference map.

**Comment by sanzinger** [comment]
Regarding patching: My idea, to specify the patching bytewise. So we do not have to track all the different state (either split, continous, ...). We can just express continous patching the same way as split patching at the same time. This would give us the flexibility to express any way of how the constant has to be patched. At the same time we can use the same 'patch instruction' to actually reverse the process and read the constant from the split instruction.

The downside is that we have slower code patching on AMD64 as we do not use the full 64 bit store but store each individual byte.

 An example implementation can be found here: https://github.com/sanzinger/graal/commit/5b979276145908617b781c8e96e592fc91a94930

**Comment by tshull226** [comment]
Hi @sanzinger. Sorry I haven't gotten a chance to look at your patch yet; I will do so soon. Yes, this can help with changes within the RuntimeCodeInstaller. However, I'm not sure if your changes will be helpful for the reference map implementation and also the implementation of reference visitors and adjusters.

**Comment by sanzinger** [comment]
I think this would help here. The implementation I have provided has following properties
* agnostic to any byte order
* agnostic to any instruction set
* variable lengths instructions
* can also verify, if the value fits into the instruction sequence
* can be used for reading the value from the patched value
* reused in the assembly process itself (see `AArch64MacroAssembler.patchJumpTarget`)
* `Pacher` can be annotated with the right ELF relocations
* Has another neat properties for serialization of metadata used in our enterprise version

**Comment by tshull226** [comment]
Hi @sanzinger @christianhaeubl @christianwimmer @peter-hofer. I'm preparing a patch to submit for this, and would like your opinion on how I should introduce my changes. Within ObjectReferenceVisitor, I need to introduce the notion of a reference split into multiple pieces. Right now I am passing an `int numPieces` parameter to convey this. Should I add a new method to ObjectReferenceVisitor which has `numPieces` as a parameter, or are you fine with me adding information to the existing `visitObjectReferenceInline(Pointer objRef, @SuppressWarnings("unused") int innerOffset, boolean compressed)` method

**Comment by tshull226** [comment]
Hi. I just added a branch (https://github.com/tshull226/graal/tree/aarch64_runtime_inline_object_code_installation) which has my current implementation of the inline object code installation support. I'm made many changes since i've verified this works on AArch64, including the new notion of shiftedOffsets, so I'll wait to make a PR until I am more confident what i have is correct. However, I figured I should let you know in case you have comments in the meantime.
